src/ConnectedMode/QualityProfiles/QualityProfileDownloader.cs

```diff
@@ -94,19 +93,19 @@ public async Task<bool> UpdateAsync(BoundSonarQubeProject boundProject, IProgres
 
             EnsureProfilesExistForAllSupportedLanguages(boundProject);
 
+            var qpToUpdate = await outOfDateQualityProfileFinder.GetAsync(boundProject, cancellationToken);
+            
             var bindingConfigs = new List<IBindingConfig>();
 
-            var languageCount = languagesToBind.Count();
             int currentLanguage = 0;
+            var totalLanguages = languagesToBind.Count();
 
-            foreach (var language in languagesToBind)
+            foreach (var (language, qualityProfileInfo) in qpToUpdate)
             {
                 currentLanguage++;
 
                 var progressMessage = string.Format(BindingStrings.DownloadingQualityProfileProgressMessage, language.Name);
-                progress?.Report(new FixedStepsProgress(progressMessage, currentLanguage, languageCount));
-
-                var qualityProfileInfo = await TryDownloadQualityProfileAsync(boundProject, language, cancellationToken);
+                progress?.Report(new FixedStepsProgress(progressMessage, currentLanguage, totalLanguages));
 
                 if (qualityProfileInfo == null)
```

Comment:
Can `qualityProfileInfo` ever be `null` now? Would that be a bug in the `OutOfDateQualityProfileFinder`?

If so, then this check and the associated test could be removed, and perhaps replaced with `Debug.Assert` instead (or by an assertion earlier that checks that none of the entries in `qpToUpdate` are `null`).

Reply:
You're right, it should never be null. Debug.Assert is not needed. Also, there were no tests that covered this path.